core/Updates/5.1.0-b5.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link    https://matomo.org
+ * @license https://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Updates;
+
+use Piwik\Updater;
+use Piwik\Updates as PiwikUpdates;
+use Piwik\Updater\Migration\Factory as MigrationFactory;
+
+class Updates_5_1_0_b5 extends PiwikUpdates
+{
+    /**
+     * @var MigrationFactory
+     */
+    private $migration;
+
+    public function __construct(MigrationFactory $factory)
+    {
+        $this->migration = $factory;
+    }
+
+    public function getMigrations(Updater $updater)
+    {
+        $migration1 = $this->migration->plugin->activate('FeatureFlags');
```

Comment:
Thanks for spotting this, this was a really fun one to debug.

**Problem 1**

I made a new version and ran through the update, I see the problem.

Because this plugin is defined in `Plugin\Manager:$pluginToAlwaysActivate` it then thinks 'its always active'
See the check here: `isPluginActivated`, which means the migration doesn't do anything because it thinks its already active (its not).

I then looked at how `CoreVue` was introduced as it is the most recent addition there, and from what I can tell. It was first migrated through the migration file, then released. And then a subsequent release was done to make it so it can't be deactivated, I think thats the way.

Here is the [commit/line](https://github.com/matomo-org/matomo/commit/2ac2bc1aeaf8efce6bb9af92506311da99e1757f#diff-05382afe1086fa10ba46fd4ee095e31584c23969091ad4dc40979077f2b6fe30R41) where CoreVue migration was added.
Here is the [release](https://github.com/matomo-org/matomo/releases/tag/4.6.0-b1) its in.

Then [a follow up commit here](https://github.com/matomo-org/matomo/commit/88cbc95b7c96e13d67cc35b88f43f8e6c81dc571) to add it to `pluginToAlwaysActivate`
[Which is in this later release
](https://github.com/matomo-org/matomo/releases/tag/4.6.0-b3)

**Problem 2**
Somewhere deep in the plugin activation it checks if the plugin 'looks like a plugin' so it either has a file with its name in the root of the plugin directory, or it has a `plugin.json`, so I've just added a plugin.json to make it happy.


Solution:
Remove from `pluginToAlwaysActivate` for now, and i've added a `plugin.json`

Reply:
You know what is "nice"? Without DI it seems we don't need the plugin to be activated to use feature flags :sweat_smile: 

I was able to just deactivate the plugin and still had the example feature flag output. No console commands available, but the feature flag itself worked.

Unless anyone has an idea how to have it in `pluginsToAlwaysActivate` without the mentioned problems, perhaps a check using `PluginManager::isPluginActivated()` to deactivate any feature flags by default if the plugin is deactivated? Feels weird otherwise to be able to deactivate the plugin and still have it working.

Reply:
This plugin will always be active, especially if the feature flag managers get put in the code via DI this will break things if it gets deactivated.

I know in this PR I did I'm *not* using the DI in the home controller, that's just to avoid any risk of it not working due to the plugin not being activated initially.

Reply:
Plugins should always have a plugin class named the same as the plugin itself. Even if the class is empty.
And plugins bundled with core should avoid having there own plugin.json file, unless they should have their own versioning for some reason.

Reply:
Thanks @sgiehl I've done that, is the reason just for consistency ?